rfd/0105-ec2-tags.md

```diff
@@ -0,0 +1,119 @@
+---
+authors: Roman Tkachenko (roman@goteleport.com), Scale Team
+state: draft
+---
+
+# RFD 105 - Importing EC2 tags via API
+
+## Required approvers
+
+Engineering: @fspmarshall, @espadolini, @rosstimothy
+
+## What
+
+Proposes a way to import instance tags using AWS EC2 API instead of metadata
+endpoint during EC2 auto-discovery.
+
+## Why
+
+Teleport nodes running on AWS currently automatically import EC2 instance tags
+as server labels using instance metadata endpoint, as was implemented by the
+[RFD 72](https://github.com/gravitational/teleport/blob/master/rfd/0072-ec2-tags.md).
+
+Some users disable metadata endpoint due to security concerns but still would
+like to have Teleport import server labels from AWS.
+
+This RFD describes a mechanism by which instance labels will instead be fetched
+by the discovery service and reconciled with the node's own labels.
+
+## Scope
+
+This feature will only apply to nodes enrolled in the cluster using auto-discovery
+mechanism. Other nodes, even if they're running on EC2, will keep functioning
+as-is (at least for now), i.e. importing tags from EC2 metadata endpoint if it's
+available.
+
+## How
+
+We will discuss design and implementation of 3 components of the solution:
+
+- Fetching EC2 instance tags.
+- Reconciling tags during heartbeat.
+- Propagating tags to the node.
+
+### Fetching tags
+
+Teleport discovery service is currently responsible for scanning user's AWS
+account on a regular cadence and enrolling matching instances into the cluster
+([RFD 57](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md)).
+It is a service that users run on their own infrastructure.
+
+Discovery service uses `DescribeInstances` API. Right now the returned instance
+list is only used for cluster enrollment (i.e. running appropriate commands on
+them through SSM). With the changes proposed in this RFD, the discovery service
+will convert each EC2 `Instance` into a new helper Teleport resource that we'll
+call `DiscoveredServer`. `DiscoveredServer` will include the instance's tags as
+labels, as well as other supporting information like instance ID, account ID
+and region. Each label key will be prefixed with `aws/`.
+
+After each discovery loop cycle, the discovery service will save all new and
+updated `DiscoveredServer` resources in the cluster backend. Discovery service
+maintains an in-memory cache of all discovered EC2 instances and will only
+upsert resources that are new or had their labels changed.
+
+To reduce load on auth and backend, discovery service will send them in batches
+of `len(instances) / 600` using `UpsertDiscoveredServers()` API and stagger the
+batches over some time interval (e.g. one batch per second).
+
+`DiscoveredServer` resources will have a 6 hour TTL and the discovery service
+will re-upsert the resource prior to its expiration, in addition to when the
+resource changes in the local cache.
```

Comment:
I'm not sure if there is a right answer here, but my gut feeling is that 6 hours is too long.

Some users have very high churn clusters, with tens of thousands of nodes being added and removed every day.  None of those users use discovery AFAIK, but it's probably safest to assume that some will at some point, and six hours strikes me as too long for a high-churn cluster.  I'd be hesitant to go over 2 hours at the absolute most.

FWIW, while reducing total network/backend usage is good, reducing *peak* usage is far more important.

After some deliberation, I've come up w/ what I think is a reasonable combination of parameters (@rosstimothy I'd appreciate you double-checking my thinking here)...

- Use a TTL of 90m and jitter it a bit so that we spread out subsequent "heartbeats".
- Use a more conservative batch size of `len(instances) / 900` (targets 15m to heartbeat of all instances rather than 10m), but compensate for this longer worst-case by using a *minimum* batch size of 5 or 6 so that smaller (1-3k) clusters can get their labels propagated faster.
- Re-queue resources for "heartbeat" when they are 30m from expiring (i.e. shoot for a "heartbeat interval" of approximately 1h, but with a safety margin of 2x our expected worst-case time for writing all batches if heartbeat jitters sync up perfectly).


The above parameters were picked kind of at random, but I was roughly targeting these constraints for a 50K cluster w/ 100% daily churn:

- Less then 2% increase in average backend writes.
- Less than 10% increase in peak/worst-case backend writes.
- Less than 10% of unexpired `DiscoveredServer` resources being "orphan" resources at a given time.

Reply:
@fspmarshall When you say batch size `len(instances) / 600` - are you talking about _batch size_ or _interval between batches_?

I originally interpreted it as the minimum batch size is 600. Or did you mean this as in "spread out all discovered nodes evenly over the 10 minute interval"?

If it's the latter, I think this will result in a significant UX degradation for small clusters if you have to wait 15 minutes for your 100-node cluster labels to propagate. I agree we need to have a minimum batch size. 5-6 sounds a little bit too low though, any reason you picked these numbers and not, say, 10 or 20?

If we talk specifically about the algorithm:

- We get all EC2 instances. Let's say, we have `N` of them.
- Now we determine our batch size. We divide `N` by our minimum batch size, say 10.
  - If the result is <= 900, means we can send them all in minimum batches over 15 minutes. We send them 10 per second. 100 node cluster will have labels propagated within 10 seconds (give or take), and we'll support up to 9K clusters with minimum batch size.
  - If the result is > 900, means we need to increase the batch size so we can send them all within 15 minutes. We send them in `N/900` batches every second. I.e. if we have 9001 nodes, one of batches will be 11 instead of 10, etc.

Is this what you had in mind?

Reply:
> I originally interpreted it as the minimum batch size is 600. Or did you mean this as in "spread out all discovered nodes evenly over the 10 minute interval"?

I meant the latter.  Assuming ~1s send interval, break up the instance set to be heartbeat over the course of 10m (or 15m or whatever the chosen "worst case time" is).
 
> If it's the latter, I think this will result in a significant UX degradation for small clusters if you have to wait 15 minutes for your 100-node cluster labels to propagate. I agree we need to have a minimum batch size. 5-6 sounds a little bit too low though, any reason you picked these numbers and not, say, 10 or 20?

The exact number is arbitrary, but I have various reasons for why I'm trying to err on the side of caution here:

- I strongly prefer us under-provisioning capacity to begin with and holding open the option to scale up capacity in the future, rather than hitting critical issues and need to push out an emergency patch to downgrade capacity (thereby causing a degradation in an experience for which users have now established expectations).

- Precedence is a real concern, since we don't often only use a pattern once.  Minimizing backend load and network usage is a big part of improving teleport scalability.  Teleport already regularly fails to meet our desired scalability targets.  The more we move the goalposts on how much additional load it is acceptable for a new feature to add, the harder it will be to even maintain teleport's current level of scalability, let alone increase it.  And no matter what number we pick, the change in load characteristics being caused by this feature is pretty unprecedented given how narrow the feature's functional scope is.

- Backend scaling issues tend to be difficult to test for/predict, and disproportionately affect our biggest users when they do show up.  Backend scaling issues caused by an optional feature are even more likely to slip through the cracks and show up unexpectedly.

- RE: UX specifically... IMO if we're not confident/comfortable with slow propagation, we shouldn't implement this feature at all.  I think this only works if we accept that it's a lazy/low-priority operation, which might be made even lazier/lower priority in the future if scaling concerns demand it (or be reworked in other ways we have yet to think up... e.g. disallowing sending until we're out of cache init phase).  I think trying to commit to providing an experience indistinguishable or nearly indistinguishable from a local metadata endpoint just isn't feasible with how teleport currently works.

> We send them in N/900 batches every second. I.e. if we have 9001 nodes, one of batches will be 11 instead of 10, etc.

Probably simpler to round the dynamic batch size up (e.g. `(len(instances) / 900) + 1`).

> Is this what you had in mind?

Yeah, that's basically it. Oversimplification that omits caching/bounds checks/etc:

```golang
    instances := getInstances()
    batchSize := defaults.MinBatchSize
    if dynamicBatchSize := (len(instances) / 900) + 1; dynamicBatchSize > batchSize {
        batchSize = dynamicBatchSize
    }
    for len(instances) > 0 {
        send(instances[:batchSize])
        instances = instances[batchSize:]
        <-interval.Next()
    }
```

Reply:
Lazy propagation is fine, I think the algorithm we discussed above provides a good compromise between UX and scalability. I've pushed an update to document it in the RFD, can you guys take another look?